### PR TITLE
refactor(connector): add amount conversion framework to cybersource

### LINKED ISSUE #5942: [REFACTOR]: [CYBERSOURCE] Add amount conversion framework to Cybersource
### :memo: Feature Description

Currently, amounts are represented as `i64` values throughout the application. We want to introduce a `Unit` struct that explicitly states the denomination. A new type, `MinorUnit`, has been added to standardize the flow of amounts across the application. This type will now be used by all the connector flows.

Rather than handling conversions in each connector, we will centralize the conversion logic in one place within the core of the application.

### :hammer: Possible Implementation

- For each connector, we need to create an amount conversion function. Connectors will specify the format they require, and the core framework will handle the conversion accordingly.
- Connectors should invoke the `convert` function to receive the amount in their required format.
- Refer to the [connector documentation](https://developer.cybersource.com/docs/cybs/en-us/api-fields/reference/all/so/api-fields/original-transaction-amount.html) to determine the required amount format for each connector.
- You can refer [this PR](https://github.com/juspay/hyperswitch/pull/4825) for more context.

🔖 Note: All the changes needed should be contained within `hyperswitch/crates/router/src/connector/` , `crates/router/src/types/api.rs` , `crates/router/tests/connectors/`

### :package: Have you spent some time checking if this feature request has been raised before?

- [X] I checked and didn't find a similar issue

### :package: Have you read the Contributing Guidelines?

- [X] I have read the [Contributing Guidelines](https://github.com/juspay/hyperswitch/blob/main/docs/CONTRIBUTING.md)

### :package: Submission Process:
- Ask the maintainers for assignment of the issue, you can request for assignment by commenting on the issue itself.
- Once assigned, submit a pull request (PR).
- Maintainers will review and provide feedback, if any. 
- Maintainers can unassign issues due to inactivity, [read more here](https://github.com/juspay/hyperswitch/wiki/Hacktoberfest-Contribution-Rules).

Refer [here](https://github.com/juspay/hyperswitch/blob/main/docs/TERMS_OF_CONTEST.md) for Terms and conditions for the contest.

###  :sparkles: Are you willing to submit a PR?

### PR DESCRIPTION
**Pull request by gauravghodinde**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

added StringMajorUnit for amount conversion

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

fixes #5942 

## How did you test it?

Tested through postman:

- Create a MCA for cybersource:
- Create a Payment with Cybersource:
```
{
    "amount": 499,
    "currency": "USD",
    "confirm": true,
    "capture_method": "automatic",
    "customer_id": "test_rec7",
    "email": "guest@example.com",
    "request_external_three_ds_authentication": true,
    "customer_acceptance": {
        "acceptance_type": "online"
    },
    "payment_method": "card",
    "payment_method_type": "credit",
    "payment_method_data": {
        "card": {
            "card_number": "4242424242424242",
            "card_exp_month": "03",
            "card_exp_year": "2030",
            "card_holder_name": "Test Holder",
            "card_cvc": "737"
        }
    },
    "billing": {
        "address": {
            "city": "test",
            "country": "US",
            "line1": "here",
            "line2": "there",
            "line3": "anywhere",
            "zip": "560095",
            "state": "Washington",
            "first_name": "One",
            "last_name": "Two"
        },
        "phone": {
            "number": "1234567890",
            "country_code": "+1"
        },
        "email": "guest@example.com"
    },
    "authentication_type": "no_three_ds"
}
```
- The payment should be succeeded 
```
{
    "payment_id": "pay_LGt7Tbil8tLko2OpVHqJ",
    "merchant_id": "merchant_1730121122",
    "status": "succeeded",
    "amount": 499,
    "net_amount": 499,
    "shipping_cost": null,
    "amount_capturable": 0,
    "amount_received": 499,
    "connector": "cybersource",
    "client_secret": "pay_LGt7Tbil8tLko2OpVHqJ_secret_dycae051HVaAJYXSCdCH",
    "created": "2024-10-28T13:12:22.609Z",
    "currency": "USD",
    "customer_id": "test_rec7",
    "customer": {
        "id": "test_rec7",
        "name": null,
        "email": "guest@example.com",
        "phone": null,
        "phone_country_code": null
    },
    "description": null,
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": null,
    "off_session": null,
    "capture_on": null,
    "capture_method": "automatic",
    "payment_method": "card",
    "payment_method_data": {
        "card": {
            "last4": "4242",
            "card_type": null,
            "card_network": "Visa",
            "card_issuer": null,
            "card_issuing_country": null,
            "card_isin": "424242",
            "card_extended_bin": null,
            "card_exp_month": "03",
            "card_exp_year": "2030",
            "card_holder_name": null,
            "payment_checks": {
                "avs_response": {
                    "code": "Y",
                    "codeRaw": "Y"
                },
                "card_verification": {
                    "resultCode": "M",
                    "resultCodeRaw": "M"
                }
            },
            "authentication_data": null
        },
        "billing": null
    },
    "payment_token": "token_LyziBYSQnkubuQbE91YF",
    "shipping": null,
    "billing": {
        "address": {
            "city": "test",
            "country": "US",
            "line1": "here",
            "line2": "there",
            "line3": "anywhere",
            "zip": "560095",
            "state": "Washington",
            "first_name": "One",
            "last_name": "Two"
        },
        "phone": {
            "number": "1234567890",
            "country_code": "+1"
        },
        "email": "guest@example.com"
    },
    "order_details": null,
    "email": "guest@example.com",
    "name": null,
    "phone": null,
    "return_url": null,
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": null,
    "statement_descriptor_suffix": null,
    "next_action": null,
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": "credit",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": {
        "customer_id": "test_rec7",
        "created_at": 1730121142,
        "expires": 1730124742,
        "secret": "epk_bb3a50134caa4214ad2ea7b534184d26"
    },
    "manual_retry_allowed": false,
    "connector_transaction_id": "7301211437366791304951",
    "frm_message": null,
    "metadata": null,
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": "pay_LGt7Tbil8tLko2OpVHqJ_1",
    "payment_link": null,
    "profile_id": "pro_2pDEf41cwojHtrM9xItk",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_AFVVYXejTDoWa4IgaPnw",
    "incremental_authorization_allowed": false,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-10-28T13:27:22.609Z",
    "fingerprint": null,
    "browser_info": null,
    "payment_method_id": null,
    "payment_method_status": null,
    "updated": "2024-10-28T13:12:24.692Z",
    "charges": null,
    "frm_metadata": null,
    "merchant_order_reference_id": null,
    "order_tax_amount": null,
    "connector_mandate_id": null
}
```
- Check cybersource dashboard with respective connector_transaction_id to get the transaction data
- the amount in dashboard should match with the amount passed in request.
image

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/connector/cybersource.rs [hunk 1]
```diff
@@ -384,7 +394,17 @@ impl
...
+            req.request
+                .minor_amount
+                .ok_or(errors::ConnectorError::MissingRequiredField {
+                    field_name: "minor_amount",
+                })?;
+        let amount = convert_amount(self.amount_converter, minor_amount, req.request.currency)?;
+        let connector_router_data = cybersource::CybersourceRouterData::from((amount, req));
+
+        let connector_req =
+            cybersource::CybersourceZeroMandateRequest::try_from(&connector_router_data)?;
```
Comment (deepanshu-iiitu) [inline crates/router/src/connector/cybersource.rs:407]: This change is not required because amount is always 0 for setup mandate flow. Please revert this change.

File: crates/router/src/connector/cybersource.rs [hunk 2]
```diff
@@ -718,12 +738,19 @@ impl ConnectorIntegration<api::Capture, types::PaymentsCaptureData, types::Payme
...
-            req,
-        ))?;
+        )?;
+        let connector_router_data = cybersource::CybersourceRouterData::from((amount, req));
+        // let connector_router_data = cybersource::CybersourceRouterData::try_from((
+        //     &self.get_currency_unit(),
+        //     req.request.currency,
+        //     req.request.amount_to_capture,
+        //     req,
+        // ));
```
Comment (deepanshu-iiitu) [inline crates/router/src/connector/cybersource.rs:753]: Please remove these comments.

File: crates/router/src/connector/cybersource/transformers.rs [hunk 3]
```diff
@@ -41,21 +41,30 @@ use crate::{
...
+//         (currency_unit, currency, amount, item): (&api::CurrencyUnit, enums::Currency, i64, T),
+//     ) -> Result<Self, Self::Error> {
+//         // This conversion function is used at different places in the file, if updating this, keep a check for those
+//         let amount = utils::get_amount_as_string(currency_unit, amount, currency)?;
+//         Ok(Self {
+//             amount,
+//             router_data: item,
+//         })
+//     }
+// }
```
Comment (deepanshu-iiitu) [inline crates/router/src/connector/cybersource/transformers.rs:60]: Remove these comments

File: crates/router/src/connector/cybersource/transformers.rs [hunk 4]
```diff
@@ -85,15 +94,20 @@ pub struct CybersourceZeroMandateRequest {
...
         let email = item.get_billing_email().or(item.request.get_email())?;
         let bill_to = build_bill_to(item.get_optional_billing(), email)?;
 
         let order_information = OrderInformationWithBill {
             amount_details: Amount {
-                total_amount: "0".to_string(),
+                total_amount: item_data.amount.clone(),
                 currency: item.request.currency,
             },
             bill_to: Some(bill_to),
```
Comment (deepanshu-iiitu) [inline crates/router/src/connector/cybersource/transformers.rs:113]: These changes are not required. Hard code the value to zero.
`StringMajorUnit("0".to_string())`

File: crates/router/src/connector/cybersource/transformers.rs [hunk 5]
```diff
@@ -3495,17 +3504,18 @@ pub struct CybersourceRefundRequest {
...
                 amount_details: Amount {
-                    total_amount: item.amount.clone(),
-                    currency: item.router_data.request.currency,
+                    total_amount: item_data.amount.clone(),
+                    currency: item.request.currency,
                 },
             },
             client_reference_information: ClientReferenceInformation {
-                code: Some(item.router_data.request.refund_id.clone()),
+                code: Some(item.request.refund_id.clone()),
```
Comment (deepanshu-iiitu) [inline crates/router/src/connector/cybersource/transformers.rs:3518]: These changes are not required.

File: crates/router/src/connector/cybersource/transformers.rs [hunk 6]
```diff
@@ -2759,12 +2773,13 @@ impl TryFrom<&CybersourceRouterData<&types::PaymentsPreProcessingRouterData>>
...
-        item: &CybersourceRouterData<&types::PaymentsPreProcessingRouterData>,
+        item_data: &CybersourceRouterData<&types::PaymentsPreProcessingRouterData>,
     ) -> Result<Self, Self::Error> {
+        let item = item_data.router_data.clone();
         let client_reference_information = ClientReferenceInformation {
-            code: Some(item.router_data.connector_request_reference_id.clone()),
+            code: Some(item.connector_request_reference_id.clone()),
         };
-        let payment_method_data = item.router_data.request.payment_method_data.clone().ok_or(
+        let payment_method_data = item.request.payment_method_data.clone().ok_or(
```
Comment (deepanshu-iiitu) [inline crates/router/src/connector/cybersource/transformers.rs:2782]: These changes are not required. please revert them

File: crates/router/src/connector/cybersource/transformers.rs [hunk 7]
```diff
@@ -2814,15 +2829,15 @@ impl TryFrom<&CybersourceRouterData<&types::PaymentsPreProcessingRouterData>>
...
             errors::ConnectorError::MissingRequiredField {
                 field_name: "redirect_response",
             },
         )?;
 
         let amount_details = Amount {
-            total_amount: item.amount.clone(),
-            currency: item.router_data.request.currency.ok_or(
+            total_amount: item_data.amount.clone(),
+            currency: item.request.currency.ok_or(
```
Comment (deepanshu-iiitu) [inline crates/router/src/connector/cybersource/transformers.rs:2840]: These changes are not required. please revert them

File: crates/router/src/connector/cybersource/transformers.rs [hunk 8]
```diff
@@ -2855,10 +2867,7 @@ impl TryFrom<&CybersourceRouterData<&types::PaymentsPreProcessingRouterData>>
                         client_reference_information,
                         consumer_authentication_information:
                             CybersourceConsumerAuthInformationRequest {
-                                return_url: item
-                                    .router_data
-                                    .request
-                                    .get_complete_authorize_url()?,
+                                return_url: item.request.get_complete_authorize_url()?,
```
Comment (deepanshu-iiitu) [inline crates/router/src/connector/cybersource/transformers.rs:2870]: These changes are not required. please revert them

**Review by deepanshu-iiitu — CHANGES_REQUESTED** [review]
Please implement the suggested changes.

**Comment by deepanshu-iiitu** [comment]
@gauravghodinde Please implement the suggested changes

**Comment by gauravghodinde** [comment]
@deepanshu-iiitu  please check this out, https://github.com/juspay/hyperswitch/pull/6335#discussion_r1805420761

File: crates/router/src/connector/cybersource/transformers.rs (hunk 4 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/cybersource/transformers.rs:113]: You can make the function public

Comment (gauravghodinde) [reply crates/router/src/connector/cybersource/transformers.rs:113]: sure making the new function public

**Comment by gauravghodinde** [comment]
hey @deepanshu-iiitu, i please review i have made the requested changes

**Review by deepanshu-iiitu — APPROVED** [review]

File: crates/common_utils/src/types.rs [hunk 9]
```diff
@@ -581,6 +581,11 @@ impl StringMajorUnit {
         Ok(MinorUnit::new(amount_i64))
     }
 
+    /// forms a new StringMajorUnit default unit i.e zero
+    pub fn zero() -> Self {
+        Self("0".to_string())
+    }
+
```
Comment (dracarys18) [inline crates/common_utils/src/types.rs:588]: Why do we need this? You can call StringMajorUnit::new(String::new("0"))

**Review by deepanshu-iiitu — DISMISSED** [review]

**Comment by deepanshu-iiitu** [comment]
I've tested the Payin flows, and everything is working as expected. We can proceed with the merge once @Sakilmostak completes the testing on the payouts side and provides his approval.

**Review by Sakilmostak — APPROVED** [review]

**Review by dracarys18 — APPROVED** [review]

**Comment by Gnanasundari24** [comment]
@deepanshu-iiitu Can you add the testcases for cybersource

**Comment by gauravghodinde** [comment]
hey @deepanshu-iiitu  can you explain why the checks failed?

File: crates/router/src/connector/cybersource/transformers.rs [hunk 10]
```diff
@@ -93,7 +88,7 @@ impl TryFrom<&types::SetupMandateRouterData> for CybersourceZeroMandateRequest {
 
         let order_information = OrderInformationWithBill {
             amount_details: Amount {
-                total_amount: "0".to_string(),
+                total_amount: StringMajorUnit::new("0".to_string()),
```
Comment (SanchithHegde) [inline crates/router/src/connector/cybersource/transformers.rs:91]:
```suggestion
                total_amount: StringMajorUnit::zero(),
```

And can you also make `StringMajorUnit::new()` private for now?

Comment (gauravghodinde) [reply crates/router/src/connector/cybersource/transformers.rs:91]: sure

**Review by SanchithHegde — APPROVED** [review]

**Review by deepanshu-iiitu — APPROVED** [review]
